File: client/src/main/java/com/jraska/github/client/TopActivityProvider.java
package com.jraska.github.client;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;
import android.support.annotation.NonNull;
import com.jraska.github.client.ui.BaseActivity;

import javax.inject.Provider;

public class TopActivityProvider implements Provider<BaseActivity> {
  private BaseActivity topActivity;

  final Application.ActivityLifecycleCallbacks callbacks = new Application.ActivityLifecycleCallbacks() {
    @Override public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
      topActivity = (BaseActivity) activity;
    }

    @Override public void onActivityStarted(Activity activity) {
    }

    @Override public void onActivityResumed(Activity activity) {
      topActivity = (BaseActivity) activity;
    }

    @Override public void onActivityPaused(Activity activity) {
    }

    @Override public void onActivityStopped(Activity activity) {
    }

    @Override public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
    }

    @Override public void onActivityDestroyed(Activity activity) {
    }
  };

  TopActivityProvider() {
  }

  @NonNull
  @Override
  public BaseActivity get() {
    if (topActivity == null) {
      throw new IllegalStateException("No activity");
    }

    return topActivity;
  }
}

File: client/src/main/java/com/jraska/github/client/TopActivityProvider.kt
package com.jraska.github.client

import android.app.Activity
import android.app.Application
import android.os.Bundle
import com.jraska.github.client.ui.BaseActivity

import javax.inject.Provider

class TopActivityProvider internal constructor() : Provider<BaseActivity> {
  private var topActivity: BaseActivity? = null

  val callbacks: Application.ActivityLifecycleCallbacks = object : Application.ActivityLifecycleCallbacks {
    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
      topActivity = activity as BaseActivity
    }

    override fun onActivityStarted(activity: Activity) {}

    override fun onActivityResumed(activity: Activity) {
      topActivity = activity as BaseActivity
    }

    override fun onActivityPaused(activity: Activity) {}

    override fun onActivityStopped(activity: Activity) {}

    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle?) {}

    override fun onActivityDestroyed(activity: Activity) {}
  }

  override fun get(): BaseActivity {
    if (topActivity == null) {
      throw IllegalStateException("No activity")
    }

    return topActivity!!
  }
}
